### ISSUE TITLE
The OUGC statistics does not match forum statistics

### ISSUE THREAD

**Issue by AlekzNet** [Sama34/OUGC-Chart-Stats#3]
Just an example for 1 day:
Real new users: -2
OUGC stats new users: 33
Real new posts: 10
OUGC stats new posts: 61
Real active users: 209
OUGC stats active users: 27

**Comment by Sama34** [comment]
> Real new users: -2

How come?

Where are you taking the _real_ stats from?

**Comment by AlekzNet** [comment]
> How come?

Apparently the users did not confirm the email addresses and were deleted by the cleaning up scripts.

> Where are you taking the _real_ stats from?

From the forum stats underneath the index, from the Statistics page in the admin section, or by simply counting new posts: 61 and 10 = the difference is way too big, even if the countings are made at different times of the day. So, I think, that something fundamental in how the plugin counts the differences is not quite right.

**Comment by AlekzNet** [comment]
Oh, interesting, the plugin stats are different today and now they do match the forum stats! The caching is set to 1 hour, but it took half a day to settle. Also, resetting to the default 30 days was required.

**Comment by AlekzNet** [comment]
A side question: what is "Active users"? Is it the amount of online registered users at the moment of plugin stats recalculating for the day (one moment in 24hours)? Then it has no real meaning. I think it's more useful to record the amount of the "Members who Were Online Today".  What do you think?

**Comment by AlekzNet** [comment]
Even more interesting. The stats, calculated by the plugin for the same day, change over time! How come?
![plugin_stat-01](https://user-images.githubusercontent.com/10780668/85078892-c6c9a980-b1c5-11ea-850b-a4cc522a9a02.png)
![plugin_stat-02](https://user-images.githubusercontent.com/10780668/85078894-c7624000-b1c5-11ea-9ee8-a5f06ba169de.png)

**Comment by AlekzNet** [comment]
Can the stats, instead of counting posts/users every caching period, be taken from mybb_stats table?

**Comment by AlekzNet** [comment]
And the "active users" from `online.php?action=today`?

**Comment by Sama34** [comment]
> The caching is set to 1 hour, but it took half a day to settle.

It might be the case that the stats doesn't match between updates but the cache system should work by itself. I will assume the caching system is fine and the issue to be different.

> Also, resetting to the default 30 days was required.

What do you mean? What breaks if you use a different value?

> I think it's more useful to record the amount of the "Members who Were Online Today". What do you think?

There is no way to accurately display online members per day historically, instead it calculates the amount of users whose last visit was `n` day. Taking this into account, you could say that the chart displays the amount of "online" users accurately only for the current day (June 15 and 17 in the images you posted above I suppose).

> The stats, calculated by the plugin for the same day, change over time! How come?

By the explanation above only the `active users` stat should change over time using the official release package.

If you are using the develop branch then thread and post stats could also start changing because the plugin now only counts visible content regardless of who sees the stats. This is closes to MyBB style.

> Can the stats, instead of counting posts/users every caching period, be taken from mybb_stats table?

MyBB doesn't store such stats, they are basically general stats. I will check how MyBB builds the ACP statics data to compare (`Home » Statistics » Overall Statistics`).

**Comment by AlekzNet** [comment]
> > Also, resetting to the default 30 days was required.
> 
> What do you mean? What breaks if you use a different value?

The plugin stats were not close to the forum stats unless I rest the plugin settings to 30 days.

> > The stats, calculated by the plugin for the same day, change over time! How come?
> 
> By the explanation above only the `active users` stat should change over time using the official release package.

The amounts of posts and threads are also different.

> There is no way to accurately display online members per day historically, instead it calculates the amount of users whose last visit was `n` day.

Yes, this is how `online.php?action=today` calculates them. I think it would be nice if the amount of active users per day were also recorded in mybb_stats by MyBB core. Wishful thinking... A separate plugin maybe?



> Taking this into account, you could say that the chart displays the amount of "online" users accurately only for the current day (June 15 and 17 in the images you posted above I suppose).

Unfortunately the amount of "online" users is not reported accurately by the plugin. The real amount as reported by online.php is much higher. "Online today" is around 250-280. "Online now" is usually never less than 25.


> > Can the stats, instead of counting posts/users every caching period, be taken from mybb_stats table?
> 
> MyBB doesn't store such stats, they are basically general stats. I will check how MyBB builds the ACP statics data to compare (`Home » Statistics » Overall Statistics`).

`mybb_stats` saves this:
```txt
dateline | numusers | numthreads | numposts |  
```
So, what the graph requires for particular day `n` is the difference between day `n` and `n-1`.

**Comment by Sama34** [comment]
> A separate plugin maybe?

> So, what the graph requires for particular day n is the difference between day n and n-1.

Maybe, or a new feature for this plugin, but not a priority right now.

> The real amount as reported by online.php is much higher.

For users (registered and online users) this plugin only accounts for registered (existing) users. MyBB might account for any user (guests and maybe bots).

I will take the time to find issues with the current code, but new features are not a priority.

**Comment by AlekzNet** [comment]
> > The real amount as reported by online.php is much higher.
> 
> For users (registered and online users) this plugin only accounts for registered (existing) users. MyBB might account for any user (guests and maybe bots).

Yes, I checked only registered users (`online.php?action=today`).
